remill/BC/Annotate.h

```diff
@@ -0,0 +1,382 @@
+  return true;
+}
+
+// Give function OriginType
+template<typename OriginType>
+void Annotate(llvm::Function *func) {
+  auto &C = func->getContext();
+  DLOG(INFO) << "Annotating: " << func->getName().str() << ": " << OriginType::metadata_kind
+             << " -> " << OriginType::metadata_value;
+  auto node = llvm::MDNode::get(C, llvm::MDString::get(C, OriginType::metadata_value));
+  func->setMetadata(OriginType::metadata_kind, node);
+}
+
+template<typename OriginType>
+bool HasOriginType(llvm::Function *func) {
+  return GetNode<OriginType>(func);
+}
+
+template<typename Type, typename Second, typename ...OriginTypes>
+bool HasOriginType(llvm::Function *func) {
+  return HasOriginType<Type>(func) || HasOriginType<Second, OriginTypes...>(func);
+}
+
+// Return list of functions that are one of chosen OriginType
+template<typename Container, typename ... OriginTypes>
+void GetFunctionsByOrigin(llvm::Module &module, Container &result) {
+  for (auto& func : module) {
+    if (HasOriginType<OriginTypes...>(&func)) {
+      // Method that is both in std::set and std::vector
+      result.insert(result.end(), &func);
+    }
+  }
+}
+
+template<typename Container, typename ... OriginTypes>
+Container GetFunctionsByOrigin(llvm::Module &module) {
+  Container result;
+  GetFunctionsByOrigin<Container, OriginTypes...>(module, result);
+  return result;
+}
+
+template<typename OriginType>
+void ChangeOriginType(llvm::Function *func) {
+  Annotate<OriginType>(func);
+}
+
+template<typename OriginType, typename OldType>
+bool ChangeOriginType(llvm::Function *func) {
+  if (HasOriginType<OldType>(func)) {
+    ChangeOriginType<OriginType>(func);
+    return true;
+  }
+  return false;
+}
+/* Functions that "tie" together two functions via specific metada:
+ * Both of them will contain metadata of specified kind, that is the other function.
+ * This is useful for example to tie entrypoints and their corresponding sub_ functions
+ */
+
+// Get node that is holding the tie information
+static inline llvm::MDNode *GetTieNode(llvm::Function *func, const std::string &kind = TieKind) {
+  return func->getMetadata(kind);
+}
+
+static inline bool IsTied(llvm::Function *func, const std::string& kind = TieKind) {
+  return GetTieNode(func, kind);
+}
+
+// Add metadata to first with information about second
+llvm::MDNode *TieFunction(llvm::Function *first, llvm::Function *second,
+                          const std::string& kind = TieKind);
+
+std::pair<llvm::MDNode *, llvm::MDNode *>
+TieFunctions(llvm::Function *first, llvm::Function *second, const std::string &kind = TieKind);
+
+// Get function that is tied to func, or nullptr if there is no such function
+llvm::Function *GetTied(llvm::Function *func, const std::string &kind = TieKind);
+
+
+/* Filter Tied functions in meaningful way. Several overloads are prepared depending on customization
+ * required.
+ */
+template<typename BinaryPredicate>
+void GetTieMapping(
```

Comment:
`static`

Reply:
This can probably be passed to a function implemented in the `.cpp` where `pred` binds to a `std::function`, or just move this entirely to the .cpp and make it not templatized, using `std::function`. Unless I am missing something....